File: approvaliq-app/lib/documents/contradictions.ts
import type { Contradiction, ExtractedField, Severity } from "@/types";

// ---------------------------------------------------------------------------
// Field comparison rules
// ---------------------------------------------------------------------------

type CompareMode = "numeric" | "text" | "boolean" | "identifier";

interface FieldRule {
  compare: CompareMode;
  tolerancePercent?: number; // only for numeric
  severityIfDiffer: Severity;
  normalise?: (v: unknown) => string;
}

const FIELD_RULES: Record<string, FieldRule> = {
  factoryAreaSqFt: {
    compare: "numeric",
    tolerancePercent: 2,
    severityIfDiffer: "blocking",
  },
  investmentCrore: {
    compare: "numeric",
    tolerancePercent: 5,
    severityIfDiffer: "warning",
  },
  employeeCount: {
    compare: "numeric",
    tolerancePercent: 10,
    severityIfDiffer: "warning",
  },
  companyName: {
    compare: "text",
    severityIfDiffer: "informational",
    normalise: normaliseCompanyName,
  },
  factoryAddress: {
    compare: "text",
    severityIfDiffer: "warning",
    normalise: normaliseAddress,
  },
  registrationNumber: {
    compare: "identifier",
    severityIfDiffer: "blocking",
  },
  hasBoiler: {
    compare: "boolean",
    severityIfDiffer: "blocking",
  },
};

// ---------------------------------------------------------------------------
// Normalisation helpers
// ---------------------------------------------------------------------------

function normaliseCompanyName(v: unknown): string {
  return String(v)
    .toLowerCase()
    .replace(/private limited|pvt\.? ltd\.?|pvt limited/g, "pvtltd")
    .replace(/limited|ltd\.?/g, "ltd")
    .replace(/[^a-z0-9]/g, "");
}

function normaliseAddress(v: unknown): string {
  return String(v)
    .toLowerCase()
    .replace(/\b(plot|survey|s\.?no\.?|gat)\b/g, "")
    .replace(/\b(road|rd|street|st|estate)\b/g, "")
    .replace(/[^a-z0-9]/g, "");
}

// ---------------------------------------------------------------------------
// Comparison logic
// ---------------------------------------------------------------------------

function valuesConflict(
  rule: FieldRule,
  a: unknown,
  b: unknown
): boolean {
  switch (rule.compare) {
    case "numeric": {
      const na = Number(a);
      const nb = Number(b);
      if (isNaN(na) || isNaN(nb)) return false;
      const tolerance = rule.tolerancePercent ?? 0;
      const diff = Math.abs(na - nb);
      const base = Math.max(Math.abs(na), Math.abs(nb));
      if (base === 0) return false;
      return (diff / base) * 100 > tolerance;
    }
    case "text": {
      const norm = rule.normalise ?? ((x: unknown) => String(x).toLowerCase().trim());
      return norm(a) !== norm(b);
    }
    case "identifier": {
      return (
        String(a).replace(/\s/g, "").toLowerCase() !==
        String(b).replace(/\s/g, "").toLowerCase()
      );
    }
    case "boolean": {
      return Boolean(a) !== Boolean(b);
    }
  }
}

function buildRecommendedAction(field: string, label: string): string {
  switch (field) {
    case "factoryAreaSqFt":
      return "Verify the declared factory area and resubmit the corrected plan.";
    case "factoryAddress":
      return "Confirm the correct factory address across all submitted documents.";
    case "registrationNumber":
      return "Ensure the company registration number is identical on all documents.";
    case "hasBoiler":
      return "Clarify boiler installation status — declaration and plan must agree.";
    default:
      return `Verify the ${label.toLowerCase()} across all submitted documents.`;
  }
}

function buildPredictedQuery(
  field: string,
  label: string,
  documents: string[],
  values: (string | number | boolean)[]
): string {
  const docA = documents[0]?.replace(/-/g, " ") ?? "document A";
  const docB = documents[1]?.replace(/-/g, " ") ?? "document B";
  const valA = values[0];
  const valB = values[1];

  switch (field) {
    case "factoryAreaSqFt":
      return `Please clarify the difference in declared factory area between the ${docA} (${valA} sq ft) and the submitted ${docB} (${valB} sq ft).`;
    case "factoryAddress":
      return `Please clarify the discrepancy in factory address between the ${docA} ("${valA}") and the ${docB} ("${valB}").`;
    case "registrationNumber":
      return `Please confirm the company registration number — a mismatch was found between the ${docA} and the ${docB}.`;
    default:
      return `Please clarify the difference in ${label.toLowerCase()} between the ${docA} (${valA}) and the ${docB} (${valB}).`;
  }
}

// ---------------------------------------------------------------------------
// Main export
// ---------------------------------------------------------------------------

/**
 * Detects contradictions across extracted fields from multiple documents.
 * Groups fields by name, applies field-appropriate comparison rules, and
 * returns a sorted list (blocking → warning → informational).
 *
 * Also emits a warning for any field with confidence < 0.80 (separate from
 * mismatch detection).
 */
export function detectContradictions(
  fields: ExtractedField[]
): Contradiction[] {
  const contradictions: Contradiction[] = [];
  const seen = new Set<string>();

  // Group by field name
  const grouped = new Map<string, ExtractedField[]>();
  for (const f of fields) {
    if (!grouped.has(f.field)) grouped.set(f.field, []);
    grouped.get(f.field)!.push(f);
  }

  // Mismatch detection
  for (const [fieldName, entries] of grouped.entries()) {
    const rule = FIELD_RULES[fieldName];
    if (!rule) continue; // unknown field — skip

    // Compare every pair; flag on first mismatch found
    for (let i = 0; i < entries.length; i++) {
      for (let j = i + 1; j < entries.length; j++) {
        const a = entries[i];
        const b = entries[j];
        if (!valuesConflict(rule, a.value, b.value)) continue;

        const id = `contradiction-${fieldName}`;
        if (seen.has(id)) continue; // only one contradiction per field
        seen.add(id);

        const docs = [a.sourceDocument, b.sourceDocument];
        const vals = [a.value, b.value] as (string | number | boolean)[];

        contradictions.push({
          id,
          field: fieldName,
          label: a.label,
          documents: docs,
          values: vals,
          severity: rule.severityIfDiffer,
          recommendedAction: buildRecommendedAction(fieldName, a.label),
          predictedQuery: buildPredictedQuery(fieldName, a.label, docs, vals),
        });
      }
    }
  }

  // Low-confidence warnings (separate from mismatch)
  const LOW_CONFIDENCE_THRESHOLD = 0.8;
  for (const entry of fields) {
    if (entry.confidence < LOW_CONFIDENCE_THRESHOLD) {
      const id = `low-confidence-${entry.field}-${entry.sourceDocument}`;
      if (seen.has(id)) continue;
      seen.add(id);

      contradictions.push({
        id,
        field: entry.field,
        label: entry.label,
        documents: [entry.sourceDocument],
        values: [entry.value],
        severity: "warning",
        recommendedAction: `Verify the ${entry.label.toLowerCase()} in the ${entry.sourceDocument.replace(/-/g, " ")} — extraction confidence was low (${Math.round(entry.confidence * 100)}%).`,
        predictedQuery: `Please confirm the ${entry.label.toLowerCase()} recorded in the ${entry.sourceDocument.replace(/-/g, " ")}; the submitted copy was difficult to read.`,
      });
    }
  }

  // Sort: blocking → warning → informational
  const ORDER: Record<Severity, number> = {
    blocking: 0,
    warning: 1,
    informational: 2,
  };

  return contradictions.sort((a, b) => ORDER[a.severity] - ORDER[b.severity]);
}

File: services/api/app/documents/contradictions.py
"""Cross-document contradiction detector.

Groups extracted fields by name across all documents in a pack, then
applies field-appropriate comparison rules. Returns a sorted list of
Contradiction dicts (blocking → warning → informational).

No LLM, no OCR. Deterministic and offline-safe.
"""

from __future__ import annotations

from typing import Any, Literal

Severity = Literal["blocking", "warning", "informational"]

_SEVERITY_ORDER: dict[Severity, int] = {
    "blocking": 0,
    "warning": 1,
    "informational": 2,
}

# ---------------------------------------------------------------------------
# Field comparison rules
# ---------------------------------------------------------------------------

def _normalise_company_name(v: Any) -> str:
    import re
    s = str(v).lower()
    s = re.sub(r"private limited|pvt\.? ltd\.?|pvt limited", "pvtltd", s)
    s = re.sub(r"limited|ltd\.?", "ltd", s)
    s = re.sub(r"[^a-z0-9]", "", s)
    return s


def _normalise_address(v: Any) -> str:
    import re
    s = str(v).lower()
    s = re.sub(r"\b(plot|survey|s\.?no\.?|gat)\b", "", s)
    s = re.sub(r"\b(road|rd|street|st|estate)\b", "", s)
    s = re.sub(r"[^a-z0-9]", "", s)
    return s


# Each rule: (compare_mode, tolerance_pct_or_None, severity, normalise_fn_or_None)
_FIELD_RULES: dict[str, dict] = {
    "factoryAreaSqFt":    {"mode": "numeric",     "tolerance": 2.0,  "severity": "blocking",      "normalise": None},
    "investmentCrore":    {"mode": "numeric",     "tolerance": 5.0,  "severity": "warning",       "normalise": None},
    "employeeCount":      {"mode": "numeric",     "tolerance": 10.0, "severity": "warning",       "normalise": None},
    "companyName":        {"mode": "text",        "tolerance": None, "severity": "informational", "normalise": _normalise_company_name},
    "factoryAddress":     {"mode": "text",        "tolerance": None, "severity": "warning",       "normalise": _normalise_address},
    "registrationNumber": {"mode": "identifier",  "tolerance": None, "severity": "blocking",      "normalise": None},
    "hasBoiler":          {"mode": "boolean",     "tolerance": None, "severity": "blocking",      "normalise": None},
}

_LOW_CONFIDENCE_THRESHOLD = 0.80


# ---------------------------------------------------------------------------
# Comparison helpers
# ---------------------------------------------------------------------------

def _values_conflict(rule: dict, a: Any, b: Any) -> bool:
    mode = rule["mode"]
    if mode == "numeric":
        try:
            na, nb = float(a), float(b)
        except (TypeError, ValueError):
            return False
        base = max(abs(na), abs(nb))
        if base == 0:
            return False
        return (abs(na - nb) / base * 100) > rule["tolerance"]
    if mode == "text":
        norm = rule["normalise"] or (lambda x: str(x).lower().strip())
        return norm(a) != norm(b)
    if mode == "identifier":
        return str(a).replace(" ", "").lower() != str(b).replace(" ", "").lower()
    if mode == "boolean":
        return bool(a) != bool(b)
    return False


def _recommended_action(field: str, label: str) -> str:
    actions = {
        "factoryAreaSqFt":    "Verify the declared factory area and resubmit the corrected plan.",
        "factoryAddress":     "Confirm the correct factory address across all submitted documents.",
        "registrationNumber": "Ensure the company registration number is identical on all documents.",
        "hasBoiler":          "Clarify boiler installation status — declaration and plan must agree.",
    }
    return actions.get(field, f"Verify the {label.lower()} across all submitted documents.")


def _predicted_query(field: str, label: str, doc_a: str, doc_b: str, val_a: Any, val_b: Any) -> str:
    doc_a_r = doc_a.replace("-", " ")
    doc_b_r = doc_b.replace("-", " ")
    if field == "factoryAreaSqFt":
        return (
            f"Please clarify the difference in declared factory area between the "
            f"{doc_a_r} ({val_a} sq ft) and the submitted {doc_b_r} ({val_b} sq ft)."
        )
    if field == "factoryAddress":
        return (
            f"Please clarify the discrepancy in factory address between the "
            f'{doc_a_r} ("{val_a}") and the {doc_b_r} ("{val_b}").'
        )
    if field == "registrationNumber":
        return (
            f"Please confirm the company registration number — a mismatch was found "
            f"between the {doc_a_r} and the {doc_b_r}."
        )
    return (
        f"Please clarify the difference in {label.lower()} between the "
        f"{doc_a_r} ({val_a}) and the {doc_b_r} ({val_b})."
    )


# ---------------------------------------------------------------------------
# Main export
# ---------------------------------------------------------------------------

def detect_contradictions(documents: list[dict]) -> list[dict]:
    """Detect contradictions across extracted fields from multiple documents.

    Args:
        documents: list of loaded fixture dicts (each has a ``fields`` dict).

    Returns:
        List of contradiction dicts sorted blocking → warning → informational.
    """
    contradictions: list[dict] = []
    seen: set[str] = set()

    # --- Mismatch detection -------------------------------------------------
    # Collect all (field_name, value, confidence, doc_id) tuples
    entries: dict[str, list[dict]] = {}
    for doc in documents:
        doc_id = doc.get("document_id", "unknown")
        for field_name, field_data in doc.get("fields", {}).items():
            # Skip non-dict entries (e.g. metadata strings accidentally in fields)
            if not isinstance(field_data, dict):
                continue
            entries.setdefault(field_name, []).append({
                "doc_id": doc_id,
                "label": field_name,  # will use field_name as label fallback
                "value": field_data.get("value"),
                "confidence": field_data.get("confidence", 1.0),
            })

    for field_name, field_entries in entries.items():
        rule = _FIELD_RULES.get(field_name)
        if not rule:
            continue
        # Compare every pair; emit first mismatch per field
        for i in range(len(field_entries)):
            for j in range(i + 1, len(field_entries)):
                a = field_entries[i]
                b = field_entries[j]
                if not _values_conflict(rule, a["value"], b["value"]):
                    continue
                cid = f"contradiction-{field_name}"
                if cid in seen:
                    continue
                seen.add(cid)
                contradictions.append({
                    "id": cid,
                    "field": field_name,
                    "label": field_name,
                    "documents": [a["doc_id"], b["doc_id"]],
                    "values": [a["value"], b["value"]],
                    "severity": rule["severity"],
                    "recommended_action": _recommended_action(field_name, field_name),
                    "predicted_query": _predicted_query(
                        field_name, field_name,
                        a["doc_id"], b["doc_id"],
                        a["value"], b["value"],
                    ),
                })

    # --- Low-confidence warnings --------------------------------------------
    for doc in documents:
        doc_id = doc.get("document_id", "unknown")
        for field_name, field_data in doc.get("fields", {}).items():
            if not isinstance(field_data, dict):
                continue
            conf = field_data.get("confidence", 1.0)
            if conf < _LOW_CONFIDENCE_THRESHOLD:
                cid = f"low-confidence-{field_name}-{doc_id}"
                if cid in seen:
                    continue
                seen.add(cid)
                contradictions.append({
                    "id": cid,
                    "field": field_name,
                    "label": field_name,
                    "documents": [doc_id],
                    "values": [field_data.get("value")],
                    "severity": "warning",
                    "recommended_action": (
                        f"Verify the {field_name} in the {doc_id.replace('-', ' ')} "
                        f"— extraction confidence was low ({round(conf * 100)}%)."
                    ),
                    "predicted_query": (
                        f"Please confirm the {field_name} recorded in the "
                        f"{doc_id.replace('-', ' ')}; the submitted copy was difficult to read."
                    ),
                })

    contradictions.sort(key=lambda c: _SEVERITY_ORDER.get(c["severity"], 99))
    return contradictions
